### ISSUE TITLE
python 3

### ISSUE THREAD

**Issue by bibi21000** [OpenZWave/python-openzwave#3]
I've made some updates on the lib and add some tests in tests/lib to help

**Comment by deisi** [comment]
At the HomeAssitant project, we try to use python-openzwave, but I get

```
    15-11-22 22:47:37 root: Z-Wave Notification DriverReady : ['Traceback (most recent call last):\n', '  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/network.py", line 855, in _handle_driver_ready\n    logging.debug(\'Network %s\' % self )\n', '  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/network.py", line 297, in __str__\n    (self.home_id_str,  self.controller)\n', '  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/controller.py", line 139, in __str__\n    (self._network.home_id_str, self._object_id, self._node.name, self._node.product_name, self.capabilities, self.library_description)\n', '  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/node.py", line 91, in name\n    return self._network.manager.getNodeName(self.home_id, self.object_id).decode("UTF-8")\n', "AttributeError: 'str' object has no attribute 'decode'\n"]
15-11-22 22:47:37 homeassistant.components.sensor: Error while setting up platform zwave
Traceback (most recent call last):

  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/helpers/entity_component.py", line 136, in _setup_platform
    self.hass, platform_config, self.add_entities, discovery_info)
  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/components/sensor/zwave.py", line 51, in setup_platform
    specific_sensor_key = (value.node.manufacturer_id,
  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/node.py", line 459, in manufacturer_id
    return self._network.manager.getNodeManufacturerId(self.home_id, self.object_id).decode("UTF-8")
AttributeError: 'str' object has no attribute 'decode'
15-11-22 22:47:37 homeassistant.components.switch: Error while setting up platform zwave
Traceback (most recent call last):

  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/helpers/entity_component.py", line 136, in _setup_platform
    self.hass, platform_config, self.add_entities, discovery_info)
  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/components/switch/zwave.py", line 31, in setup_platform
    add_devices([ZwaveSwitch(value)])
  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/helpers/entity_component.py", line 70, in add_entities
    self.entity_id_format, entity.name,
  File "/usr/lib/python3.5/site-packages/homeassistant-0.9.0.dev0-py3.5.egg/homeassistant/components/switch/zwave.py", line 61, in name
    name = self._node.name or "{}".format(self._node.product_name)
  File "/usr/lib/python3.5/site-packages/python_openzwave_api-0.2.6-py3.5.egg/openzwave/node.py", line 91, in name
    return self._network.manager.getNodeName(self.home_id, self.object_id).decode("UTF-8")
AttributeError: 'str' object has no attribute 'decode'
```

I've been told that this is a python-openzwave error. Any idear how to fix this? Is it already known?
[here](https://github.com/balloob/home-assistant/issues/382) is a link to the issue in HomeAssistant.

**Comment by durin42** [comment]
Experimentally, the problem is that in Python 3 the getNodeName et al methods are returning str, which is the old unicode type. It'd probably make the most sense to figure out how to get the unicode type out of python-openzwave all the time.

**Comment by deisi** [comment]
If I see this correctly, this is an issue only in python 3.5. Using python 3.4 everything works for me.

**Comment by durin42** [comment]
That's puzzling, as it looks like the Cython (libopenzwave.pyx:2098 or so) should return a bytes everywhere (based on the documentation for Cython anyway). What version of Cython are you using? What specific Python version?

I'd be happy to try and send a PR to fix this, but I'm not really sure what the variable that matters is yet. Is the intent for these methods to return unicode strings or byte strings?

**Comment by joshughes** [comment]
https://github.com/OpenZWave/python-openzwave/issues/28

I am seeing something that seems similar...  although the way I am compiling I don't think I am using Cython since I am using https://github.com/OpenZWave/python-openzwave/blob/master/INSTALL_ARCH.txt method of installation.

**Comment by partofthething** [comment]
This is a dupe of #10 at this point, where I think the issue is mostly figured out. I recommend closing this as a dupe of #10.

**Comment by durin42** [comment]
@partofthething I really don't think so. I built from your 0e708b02dd44ec5f6b58b276b971f377c92f4d61, and I'm still getting similar bytes/string problems:

```
ERROR:homeassistant.bootstrap:Error during setup of component zwave
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/homeassistant/bootstrap.py", line 98, in _setup_component
    if not component.setup(hass, config):
  File "/usr/local/lib/python3.5/site-packages/homeassistant/components/zwave.py", line 129, in setup
    DEFAULT_ZWAVE_CONFIG_PATH),)
  File "/usr/local/lib/python3.5/site-packages/pyozwweb-0.3.0b7-py3.5.egg/openzwave/option.py", line 76, in __init__
    libopenzwave.PyOptions.__init__(self, config_path=config_path, user_path=user_path, cmd_line=cmd_line)
  File "src-lib/libopenzwave/libopenzwave.pyx", line 579, in libopenzwave.PyOptions.__init__ (src-lib/libopenzwave/libopenzwave.cpp:7894)
    self.create(self._config_path, self._user_path, self._cmd_line)
  File "src-lib/libopenzwave/libopenzwave.pyx", line 582, in libopenzwave.PyOptions.create (src-lib/libopenzwave/libopenzwave.cpp:7988)
    def create(self, char *a, char *b, char *c):
TypeError: expected bytes, str found
```

Got other thoughts?

Note that if it matters, I'm doing the entire build with Python 3.5. The system in question doesn't even have a Python 2 interpreter available.

**Comment by durin42** [comment]
Looking at libopenzwave.pyx a little more, I _think_ the problem is that almost all the functions there are defined in terms of char*, which defaults to bytes on Python 3?

I'm not sure, as I've never worked with Cython before, and I'm not finding anything super-obvious in the docs.

**Comment by durin42** [comment]
https://github.com/OpenZWave/python-openzwave/pull/33 has my only-mostly-cargo-cult attempt at fixing this.

**Comment by deisi** [comment]
@durin42 using python 3.4 fixes your problem. python 3.5 doesn't work currently.

**Comment by durin42** [comment]
That's not a viable solution.

**Comment by durin42** [comment]
(Hit send too soon)

My pull request should fix 3.5 support (works here), and as far as I
understand things it should work on 3.4 as well.

**Comment by bibi21000** [comment]
Should be fixed
